src/sink/rest.rs

```diff
@@ -449,7 +456,12 @@ impl Sink for Rest {
             .collect::<HashMap<String, Box<dyn Codec>>>();
         let default_method = self.config.method.0;
         let endpoint = self.config.endpoint.clone();
-        let config_headers = self.config.headers.clone();
+        let mut config_headers = self.config.headers.clone();
+        // this could be a match case if other sink auth types are introduced
+        // for now clippy doesn't like equality checks disguised as match cases.
+        if self.config.auth.as_str() == "gcp" {
+            config_headers = block_on(auth::merge_gcp_headers(&config_headers))?
```

Comment:
This gets a bit complicated, so if I fail on explaining it please ask :D

Google uses OAuth, that means the bearer token has a lifetime and will eventually become stalled. If we only fetch it on initliasiation the sink would work for a while (hours, days or weeks even depending on googles config) and then suddenly throw up "invalid auth" errors. A restart will fix that but probably confuse the heck out of the operator.

To work around this OAuth2 implements multiple token types, one of them being a refresh token. The way we will need to handle this is by storing the `Token` in the struct and fetching `header_values` on every request. This will (based on the documentation from `gouth`, re-use an existing token and fetch and only renew it when required.

Reply:
I think I understand.

Here, in `init`, we're only ever refreshing this token once. Since the [authenticate_bearer](https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/gcp/auth.rs) function does request `header_values` and specifically refreshes the token, we need to do this _per request_  (like in `build_request`) instead of in `init`.

One thing I want to point out with that approach is that it appears to be inconsistent, at least I think, with the `gcp` approach we had in [`gcs.rs`](https://github.com/tremor-rs/tremor-runtime/blob/main/src/sink/gcs.rs). Namely that the sink will contain the headermap, but never re-authenticate it. I'd be happy to understand how I'm wrong there, still learning :)

I believe I understand the ask, I'll put something in and tag you specifically. I'd also like to open an issue if we agree that the `gcs.rs` could be improved in a similar way.

Reply:
Ja nearly :D we don't need to refresh once per request but every now and then when it times out ( there are [a lot of details in the RFC](https://datatracker.ietf.org/doc/html/rfc6749) but you might not need all of that :joy: ). We are somewhat lucky as in `Token` already implements all that.

`authenticate_bearer` has the problem that it creates a new token for every request, which we really don't want either :D so we kind of need to get rid of `authenticate_bearer` and make `Token` part of the struct so we can re-use the `Token` and just call `header_value` on every request  (that way we get the caching "for free") :D.

For GCS you are absolutely right! It has the same problem :) I didn't notice that on the originally PR for that :( so not your fault at all! I did already threw in #1108.  It should be fairly easy to fix too :D so if you want to feel free to claim the ticket ;) but no need to do it in this PR (it's not your problem :P)

Reply:
ah, that was the detail I missed. The token itself would need to integrate into the sink, then `header_value` would be generated every request, not a new token... well in that case, maybe this is doable? just return the token instead? lemme fiddle.

Reply:
Ja I think if you just put something like:

```
gcp_token: Option<Token>
```

into the sink, populate that on startup and then use `self.gcp_token.ok_or_else("no token)?.that_header_value_function()?` on each request it should do the trick (just with a lot nicer and more sensible code ;)

Reply:
y'all are total wizards with rust and I am a humble learner still. I really appreciate the advice, it's quite elegant. I'll put that in.